### PR TITLE
feat: sample the heatmap

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

The heatmap is slow to load and frequently confuses users.

* because we slice the data in CH query with LIMIT and OFFSET
* and because the results of that query don't map one-to-one with elements in the page

We field frequent questions from users similar "why is one of five items in the menu not showing clicks"

I added a mechanism to load more but it's still slow and also now frustrating because you have to click load more an unknown number of times.

## Changes

* samples the query from up to but not much more than 2 million rows

https://clickhouse.com/docs/en/sql-reference/statements/select/sample/#sample-n

In metabase for our team this returns the same results as an unsampled query but much more quickly.

I went for sampling by row count rather than a sample factor because locally the sampled 0.x version returned no results for me. This way round with only small amounts of team data we can query the whole dataset but cap the amount of data we'll query

TODO

- [x] a way for us to change sample rate in case it causes problems
- [ ] ~~a way for the user to change the sample rate in case they'd like to wait longer for more accurate data? (I don't think the heatmap needs to be _too_ accurate though~~
    - I don't think we need this. If it works they're already getting better data than before 
- [x] a way to quantify the impact of the change
    - we already gather metrics here http://grafana-prod-us/d/http-by-application-endpoint/http-by-application-endpoint?orgId=1&var-view=element-stats&from=now-30d&to=now&viewPanel=2 ranges from 300ms to over one minute and we are already slicing the data confusingly 

## How did you test this code?

running it locally and seeing it work, trying the query in metabase for our team

### CODE REVIEW

**Comment by neilkakkar** [comment]
very clever, makes sense, this is an ideal usecase for sampling, you rock 🚀

**Comment by pauldambra** [comment]
@yakkomajuri all the mentions of sampling made me realise it might help here.

File: posthog/models/element/sql.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 GET_ELEMENTS = """
 SELECT
-    elements_chain, count(1) as count, event as event_type
+    elements_chain, count(_sample_factor) as count, event as event_type
```
Comment (yakkomajuri) [inline posthog/models/element/sql.py:3]: I think having `_sample_factor` here doesn't change anything from having `1` and just confuses the reader?

File: posthog/api/element.py [hunk 2]
```diff
@@ -105,16 +91,10 @@ def stats(self, request: request.Request, **kwargs) -> response.Response:
                 "type": elements[2],
                 "elements": [ElementSerializer(element).data for element in chain_to_elements(elements[0])],
             }
-            for elements in result[:limit]
+            for elements in result
         ]
 
-        if paginate_response:
```
Comment (yakkomajuri) [inline posthog/api/element.py:111]: how does sampling remove the need for pagination?

**Review by yakkomajuri — APPROVED** [review]
great use of `SAMPLE n` - super cool!

Confused as to why we got rid of pagination - is this just a change we batched in here?

File: posthog/models/element/sql.py (hunk 1 above)
Comment (pauldambra) [reply posthog/models/element/sql.py:3]: Ah, yeah, I was robotically copying the docs without understanding 🙈 

But `count(_sample_factor)` and `count()` have the same results 👍

File: posthog/api/element.py (hunk 2 above)
Comment (pauldambra) [reply posthog/api/element.py:111]: I only added pagination to try and let folk that were willing to wait longer to get more data do just that. But the feedback was roughly "it's nice to see the data, but annoying to click the button"

So, I'm hoping that sampling means we can draw the heatmap without having to page. 🤔 May be safer to leave pagination in and have a larger page. Then I can experiment a little more with the impact.

File: posthog/models/element/sql.py (hunk 1 above)
Comment (pauldambra) [inline posthog/models/element/sql.py:3]:
```suggestion
    elements_chain, count() as count, event as event_type
```

**Pull request merged**